### ISSUE TITLE
Support running the relay chain node as an extra process

### ISSUE THREAD

**Issue by bkchr** [paritytech/cumulus#545]
Currently every parachain node includes a relay chain node. This relay chain node is required to get certain information about the relay chain that are important for the parachain, for example to query what is the best parachain block currently etc. While this makes it relative easy to run a parachain node, it also brings some problems like the high compilation time as we need to compile the parachain node and the relay chain node. Another, a more bigger problem is that parachain developers are required to update their collators when there is a new relay chain release that requires a timely update, because a new host function is added or something in the client code of the relay chain is fixed. So, it would be nice to have the relay chain node running as an extra process. Collator operators would just run an extra relay chain node (that could maybe even shared between multiple parachain nodes, but this is no real initial requirement!) and can freely update the relay chain node. The relay chain node itself could maybe directly bring the functionality required by Cumulus to connect to it or we provide some sort of wrapper (probably the best way for the first implementation).

This feature would be implemented in the following order:
1. Refactor all usages of the polkadot client to have them behind some common trait or maybe multiple traits. So, we should not have any reference to `polkadot-service` or `polkadot-client` in any of the "low-level" functionality of Cumulus. It should only use these interfaces to talk to the relay chain.
2.  Write an implementation of these traits for the "in-node relay chain" so that we are back to on par with the current implementation.
3. Research what is the best way to implement the inter-process communication. Maybe some sort of json rpc over https://crates.io/crates/parity-tokio-ipc or whatever.
4. Implement the wrapper and make it work. Running the relay chain as an external process should always be some sort of optional way of doing it. So, if the feature is compiled it should be enabled via some cli flag or something. 
5. ...
6. Profit :P

**Comment by xlc** [comment]
This should also make it possible to share a relaychain node between multiple parachain nodes? 

Will this make it possible to allow relaychain part and parachain part using different versions of Substrate?

**Comment by NukeManDan** [comment]
Would you be able to have remote relay-nodes that collators could connect to? If so, what requirements (latency, bandwidth, etc.) should be outlined?

**Comment by bkchr** [comment]
> This should also make it possible to share a relaychain node between multiple parachain nodes?

Yes, as written above. However, I don't see this as an initial requirement for this issue.

> Will this make it possible to allow relaychain part and parachain part using different versions of Substrate?

This is the whole point or better, make parachains collators not required to update when the relay chain requires a node update.

**Comment by bkchr** [comment]
> Would you be able to have remote relay-nodes that collators could connect to? If so, what requirements (latency, bandwidth, etc.) should be outlined?

As written in the issue, maybe. As this is not an initial requirement, I don't think we need to outline anything of that.

**Comment by bkchr** [comment]
I also just realized that this will probably be a little bit more complicated for collators, as the connection collators not only read data from the relay chain. They also get called by the overseer when they need to produce a new pov and give it back to the overseer. This would be a little bit more time critical, but should also be solvable.

**Comment by skunert** [comment]
After #963 was merged, it is now possible to start a parachain full node by passing the address to a relay chain full node. The parachain node will not internally create a relay-chain node but fetch all needed data via RPC. Be aware that we are viewing this as an experimental feature currently. 
**Note**: Collation is not supported at this time

Example command (assumes relay chain full node running locally on ws-port 9944):
`polkadot-collator --tmp --relay-chain-rpc-url "ws://localhost:9944"`

**Comment by purestaketdb** [comment]
Similar to @crystalin 's feedback - the two main use cases we are interested in are:

1. Not requiring updates to the collators when the relay chain version changes. This is not initially supported per the comments of July 2021.  

2. Allowing RPC servers to avoid running their own relay chains. However, for resilience, it would be preferable if either a set of relays could be designated OR if it would be supported/safe to run a set of relay RPCs behind a load balancer.

**Comment by skunert** [comment]
Thanks @crystalin and @purestaketdb for the feedback. Having multiple relay chain nodes to connect to is something I have thought about. I can look into it once the collation over RPC feature is ready. 
The main challenge will be to sort out subscription handling for this. Currently, we are listening to RPC subscriptions that notify us about new blocks on the relay chain. To switch relay-chain nodes on the fly, we need to gracefully continue a new subscription where the old one left off.

**Comment by skunert** [comment]
Experimental support for relay chain collators is now [merged](https://github.com/paritytech/cumulus/pull/1585). You can try this by passing the `--relay-chain-rpc-url` argument together with `--collator`. Next phase is additional testing to discover potential problems.

Road forward:
- Look into having a set of relay chain nodes as suggested

**Comment by skunert** [comment]
The points of the last comment have been addressed in https://github.com/paritytech/cumulus/pull/1880. Future enhancements can have their separate issues.
